src/main/java/org/radarcns/redcap/integration/Integrator.java

```diff
@@ -17,154 +17,50 @@
     }
 
-    /**
-     * Generates the {@link Set} of inputs that will be written in REDCap for finalising the
-     *      integration between REDCap project and Management Portal project. Using a
-     *      {@link MpClient}, the function retrieves the RADAR Subject Identifier and the Human
-     *      Readable Identifier. In the end, the function forces the REDCap integration
-     *      form /instrument status to {@link InstrumentStatus#COMPLETE}.
-     * @return {@link Set} of inputs that have to be written in REDCap.
-     */
-    @Override
-    public Set<RedCapInput> getInput() {
-        Subject subject = performSubjectUpdateOnMp(redCapInfo.getUrl(),
-                redCapInfo.getProjectId(), getRecordId());
-
-        Set<RedCapInput> set = new HashSet<>();
-
-        set.add(new IntegrationData(getRecordId(), redCapInfo.getEnrolmentEvent(),
-                IntegrationData.SUBJECT_ID_LABEL, subject.getSubjectId()));
-
-        set.add(new IntegrationData(getRecordId(), redCapInfo.getEnrolmentEvent(),
-                IntegrationData.HUMAN_READABLE_ID_LABEL, subject.getHumanReadableIdentifier()));
-
-        set.add(new IntegrationData(getRecordId(), redCapInfo.getEnrolmentEvent(),
-                RedCapManager.getStatusField(redCapInfo.getIntegrationForm()),
-                Integer.toString(InstrumentStatus.COMPLETE.getStatus())));
-
-        return set;
-    }
-
-    /**
-     * Sets the HTML form parameters needed to update the REDCap integration form / instrument.
-     *      It sets just the metadata not the real input, that is set by {@link RedCapUpdater}
-     *      while generating the {@link Request}.
-     * @param builder {@link okhttp3.FormBody.Builder} that has to be populated
-     * @return {@link okhttp3.FormBody.Builder} required to finale the HTTP request.
-     */
-    @Override
-    protected Builder setParameter(Builder builder) {
-        return builder.add("content", "record")
-                .add("format", "json")
-                .add("type", "eav")
-                .add("overwriteBehavior", "overwrite")
-                .add("returnContent", "count")
-                .add("returnFormat", "json");
+    public boolean handleDataEntryTrigger(){
+        Integer recordId = trigger.getRecord();
+        Integer projectId = redCapInfo.getProjectId();
+        String enrolmentEvent = redCapInfo.getEnrolmentEvent();
+        String integrationFrom = redCapInfo.getIntegrationForm();
+        URL url = redCapInfo.getUrl();
+        List<String> attributeKeys = redCapInfo.getAttributes().stream().map(a->a.getFieldName()).collect(Collectors.toList());
+        Map<String, String> attributes = redCapIntegrator.pullRecordAttributes(attributeKeys, recordId);
+        Subject subject = mpIntegrator.performSubjectUpdateOnMp(url, projectId, recordId, attributes);
+        return redCapIntegrator.updateRedCapIntegrationForm(subject, recordId, enrolmentEvent, integrationFrom);
```

Comment:
There is a deeper underlying race condition here. If the subject (with say id = 10) is in MP and was deleted from Redcap. Then, when we create a new subject in Redcap, it will again be created with the same id (10). In this case, the integration needs to fail, since a subject already exists with the current id in MP but does not correspond to the new record in Redcap. Previously, the integration app used to fail if a subject already exists in MP.
I think the `redCapIntegrator.updateRedCapIntegrationForm(...)` should only be called if the subject did not exist in MP and was created.  If the subject already exists, we only update its attributes but do not update back redcap form.

Reply:
You can create an ENUM in Subject class like -

```java
enum SubjectOperationStatus {
    UPDATED,
    CREATED,
    FAILED,
...
}
```
And return this status in the `mpIntegrator.performSubjectUpdateOnMp` function, so you can take action based on that.

Reply:
Okay sure.  👍

Reply:
Made the changes. I've added the operationStatus which would only update the form if a subject was created. I've also pulled the `subjectId` from REDCap to check if it matches the MP `subjectId` and fails the integration if they don't match.